### PR TITLE
[DO-NOT-MERGE][SPARK-33913][SPARK-33921][BUILD] Upgrade kafka to 2.7.0 and Upgrade Scala version to 2.12.12

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to upgrade Kafka version to 2.7.0 and Scala 2.12 patch version to 2.12.12.

### Why are the changes needed?

To catch up with latest Kafka bug fix and features: https://downloads.apache.org/kafka/2.7.0/RELEASE_NOTES.html

To catch up with bug fix and improvement of latest Scala 2.12 patch.

https://github.com/scala/scala/releases/tag/v2.12.11
https://github.com/scala/scala/releases/tag/v2.12.12

* Performance improvements in the collections library, the compiler
* Fix memory leak in runtime reflection's `TypeTag` caches
* Fix some `toX` methods that could expose the underlying mutability of a ListBuffer-generated collection
* ASM was upgraded to 7.3.1, allowing the optimizer to run on JDK 13+

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test.

### CODE REVIEW

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala
Comment (viirya) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala:577]: https://github.com/apache/kafka/pull/8931

**Review by HyukjinKwon — DISMISSED** [review]
LGTM if tests pass.

**Comment by viirya** [comment]
Yea, it's the same issue as https://github.com/embeddedkafka/embedded-kafka/issues/202. We need to bump Scala patch version to match Kafka's.

**Comment by viirya** [comment]
Created #30940 for upgrading Scala 2.12 patch version.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @viirya . Sorry, but, are you sure about Scala 2.12.12?

I commented on your original Scala 2.12.12 PR, too.
- https://github.com/apache/spark/pull/30940#issuecomment-751581096

**Comment by viirya** [comment]
> Hi, @viirya . Sorry, but, are you sure about Scala 2.12.12?
> 
> I commented on your original Scala 2.12.12 PR, too.
> 
>     * [#30940 (comment)](https://github.com/apache/spark/pull/30940#issuecomment-751581096)

Yea, I just saw it. If we don't want Scala 2.12.12, seems we cannot upgrade Kafka 2.7.0 too: https://github.com/apache/spark/pull/30940#issuecomment-751583360.

**Comment by viirya** [comment]
I'm going to see Jenkins test to verify it, although I already experimented locally.

**Comment by dongjoon-hyun** [comment]
No, @viirya . We detected the issues six month ago and made a Spark-side adaptation via SPARK-32436 on July 2020.

Please see the discussion here for the technical details and Scala side bug IDs. AFAIK, the bug is in Scala 2.12.12 and we are waiting for 2.12.13 .
- https://github.com/apache/spark/pull/29231

**Comment by dongjoon-hyun** [comment]
cc @srowen and @mridulm who were together on the original PR for Scala 2.12.12 bug fix.

**Comment by viirya** [comment]
> No, @viirya . You may not notice the bug, because we detected the issues six month ago and made a Spark-side adaptation via [SPARK-32436](https://issues.apache.org/jira/browse/SPARK-32436) on July 2020.
> 
> Please see the discussion here for the technical details and Scala side bug IDs. AFAIK, the bug is in Scala 2.12.12 and we are waiting for 2.12.13 .
> 
>     * #29231

I'm reading https://github.com/scala/bug/issues/12096. I understood Scala 2.12.12 was skipped now. I just meant if we cannot upgrade to Scala 2.12.12, then upgrading Kafka 2.7.0 is blocked by it.

Then we may also wait for next Kafka version which uses Scala 2.12.13 (if there is).

**Comment by dongjoon-hyun** [comment]
In short, the Scala compiler had a critical bug.

- [scala/bug#12096](https://github.com/scala/bug/issues/12096)
- [scala/bug#12002](https://github.com/scala/bug/issues/12002)

**Comment by dongjoon-hyun** [comment]
Got it, @viirya . For testing purpose, there is no problem.

**Comment by srowen** [comment]
Right, I think we'd just have to skip this for now.

**Comment by viirya** [comment]
Thanks for discussion.

**Comment by ijuma** [comment]
FYI, we adjusted the optimizer configs for Kafka to avoid the exact Scala patch version dependency issue: https://github.com/apache/kafka/pull/10174

It will be part of Kafka 2.8.0.

**Comment by dongjoon-hyun** [comment]
Thank you, @ijuma .

**Pull request closed without merging**
